### DISCUSSION TITLE
Error with custom query logging in NextJS 14

### DISCUSSION THREAD

**Discussion by SnowySailor** [prisma/prisma#25327] [Q&A] [answered]
### Bug description

I was following this guide for setting up a prisma instance with NextJS: https://www.prisma.io/docs/orm/more/help-and-troubleshooting/help-articles/nextjs-prisma-client-dev-practices

I wanted to add custom query logging, so I wrote this event registration:
```tsx
prisma.$on('query', (e: any) => {
  console.log(`Query: ${e.query}`);
  console.log(`Duration: ${e.duration}ms`);
});
```
However whenever I load any pages I get this error: `Error: PrismaClient is unable to run in this browser environment, or has been bundled for the browser (running in `unknown`).`

**When I remove the event registration block, the errors stop. The event handler WORKS (I see my logs in my terminal output), however the nextjs page render fails.** Contrary to the error message, I do not have any queries in any client components.

![image](https://github.com/user-attachments/assets/af7aa99e-b7b7-44d1-aeb7-61ab558894ae)

### How to reproduce

```tsx
import { PrismaClient } from '@prisma/client';

const prismaClientSingleton = () => new PrismaClient({
  log: [
    {
      emit: 'event',
      level: 'query',
    },
  ],
});

declare const globalThis: {
  prismaGlobal: ReturnType<typeof prismaClientSingleton>;
} & typeof global;

const prisma = globalThis.prismaGlobal ?? prismaClientSingleton();

prisma.$on('query', (e: any) => {
  console.log(`Query: ${e.query}`);
  console.log(`Duration: ${e.duration}ms`);
});

export default prisma;

if (process.env.NODE_ENV !== 'production') {
  globalThis.prismaGlobal = prisma;
}
```

Execute a prisma query from a server component:
```tsx
'use server';

// ...
import { getSceneInfo } from 'lib/scene';

export default async function MediaPage({ params }: Readonly<{ params: { sceneId: string } }>) {
  const sceneIdInt = parseInt(params.sceneId, 10);
  const sceneInfo = await getSceneInfo(sceneIdInt);
  if (!sceneInfo) {
    return notFound();
  }

  return (
    <SceneBody sceneInfo={sceneInfo} />
  );
}
```

`lib/scene.ts`:
```ts
import prisma from 'db';

// ...

export async function getSceneInfo(sceneId: number) {
  return prisma.scene.findUnique({
    where: { sceneId },
    select: SceneSelectQuery
  };
}
```

### Expected behavior

I get logs in the console and the page render doesn't error out

### Prisma information

N/A

### Environment & setup

- OS: Debian
- Database: Postgres
- Node.js version: 21

### Prisma Version

```
prisma                  : 5.19.1
@prisma/client          : 5.19.1
Computed binaryTarget   : linux-arm64-openssl-3.0.x
Operating System        : linux
Architecture            : arm64
Node.js                 : v21.7.3
Query Engine (Node-API) : libquery-engine 69d742ee20b815d88e17e54db4a2a7a3b30324e3 (at node_modules/@prisma/engines/libquery_engine-linux-arm64-openssl-3.0.x.so.node)
Schema Engine           : schema-engine-cli 69d742ee20b815d88e17e54db4a2a7a3b30324e3 (at node_modules/@prisma/engines/schema-engine-linux-arm64-openssl-3.0.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.19.1-2.69d742ee20b815d88e17e54db4a2a7a3b30324e3
Default Engines Hash    : 69d742ee20b815d88e17e54db4a2a7a3b30324e3
Studio                  : 0.502.0
```

**Comment by ludralph** [comment, marked as answer]
Hi @SnowySailor 👋 

Can you adjust your code to check that Prisma Client is not initialized in the browser?

```js
import { PrismaClient } from '@prisma/client';

const prismaClientSingleton = () => new PrismaClient({
  log: [
    {
      emit: 'event',
      level: 'query',
    },
  ],
});

declare const globalThis: {
  prismaGlobal: ReturnType<typeof prismaClientSingleton>;
} & typeof global;

let prisma: PrismaClient;

if (typeof window === 'undefined') {
  if (process.env.NODE_ENV === 'production') {
    prisma = prismaClientSingleton();
  } else {
    if (!globalThis.prismaGlobal) {
      globalThis.prismaGlobal = prismaClientSingleton();
    }
    prisma = globalThis.prismaGlobal;
  }

  prisma.$on('query', (e: any) => {
    console.log(`Query: ${e.query}`);
    console.log(`Duration: ${e.duration}ms`);
  });
} else {
  prisma = {} as PrismaClient; // This ensures that PrismaClient is not used in the browser
}

export default prisma;
```
The `typeof window === 'undefined'` check ensures that PrismaClient is only initialized on the server side.

**Reply by SnowySailor** [reply]
That does solve the issue, thanks. I wonder if it's just nextjs/prisma not playing nicely together when server actions are used (since prisma is technically in the import tree when you use server actions). Even though it's not _actually_ imported into the browser (since server actions are just converted into an API call) something seems to be getting confused somewhere. Maybe the prisma-nextjs integration docs should be updated to make this clear?

I had to move the event registration to right after `globalThis.prismaGlobal = prismaClientSingleton();` otherwise it was registering the handler multiple times on development.
